File: rag/advanced_rag/harness/orchestrator/query_rewriter.py
"""Query Rewriter

Query Rewriter has two jobs:
  - Phase 1: break the long request into simple, searchable questions so the
    retriever finds relevant content more accurately.
  - Phase 4: turn the Sufficient Context Agent's missing-pieces feedback into a
    NEW targeted search query (e.g. "you missed allergies -> search 'rashes' /
    'adverse events'").

This module implements the Phase 4 role for the orchestrator: it takes a
Sufficient Context Agent forward gap (what is missing + a search hint) and
rewrites it into a concrete, retrievable query. Unlike ``_try_replan``'s current
behavior — which reuses the gap text verbatim as a new claim description — this
produces a targeted query that actually names the missing entity + relation, so
the next search hits the gap instead of re-searching the same angle.

The rewriting is deliberately cheap and only invoked on replan (replan budget is
bounded), so it adds little latency.
"""

from __future__ import annotations

import logging

from rag.advanced_rag.harness.stats import in_phase
from rag.prompts.generator import PROMPT_JINJA_ENV, gen_json
from rag.prompts.template import load_prompt

_LOG = logging.getLogger(__name__)

REWRITE_PROMPT = load_prompt("sca_query_rewrite")


@in_phase("rewrite")
async def rewrite_gap_to_query(
    tools,
    question: str,
    gaps: list[tuple[str, str]],
    bridge_values: list | None = None,
    research_context: str = "",
) -> list[dict]:
    """Rewrite forward gaps into targeted search queries (multi-hop aware).

    Parameters
    ----------
    tools : RAGTools
        Must expose ``chat_mdl``.
    question : str
        The original user question (context for the rewrite).
    gaps : list[(what, search_hint)]
        Forward gaps (what the answer still needs, plus a search hint).
    bridge_values : list | None
        Already-resolved bridge values across the claims (e.g. the confirmed shows
        in a "which finale ran longest" question). The rewriter anchors the new
        search query to these, so a missing hop is re-searched with the resolved
        upstream value instead of re-deriving it.
    research_context : str
        Rendered block describing what previous rounds already tried (queries +
        outcomes) and what the current evidence pool roughly contains. Gives the
        rewriter enough information to aim at UNCOVERED angles instead of
        paraphrasing dead ones — the information-augmented counterpart of the
        react loop's in-context visibility.

    Returns
    -------
    list[dict] : ``[{"query": "..."}]`` — targeted, retrievable queries. Empty
        when the rewrite is unavailable / fails, in which case the caller falls
        back to the original gap text.
    """
    chat_mdl = getattr(tools, "chat_mdl", None)
    if chat_mdl is None or not gaps:
        return []
    gaps_text = "\n".join(f"- what: {g[0] or ''}; hint: {g[1] or ''}" for g in gaps)
    bridge_text = "\n".join(f"- {b}" for b in (bridge_values or []) if str(b).strip())
    rendered = PROMPT_JINJA_ENV.from_string(REWRITE_PROMPT).render(
        question=question or "",
        gaps=gaps_text,
        bridge_values=bridge_text,
        research_context=research_context,
    )
    try:
        result = await gen_json(rendered, "Output:\n", chat_mdl)
    except Exception as exc:  # noqa: BLE001
        _LOG.info("[QueryRewrite] failed: %s", exc)
        return []
    if not isinstance(result, dict):
        return []
    out = []
    for q in result.get("queries") or []:
        if isinstance(q, dict):
            query = str(q.get("query") or q.get("question") or "").strip()
        elif q:
            query = str(q).strip()
        else:
            continue
        if query and query not in out:
            out.append({"query": query})
    return out

File: internal/rag/advanced_rag/harness/orchestrator/query_rewriter.go
//
//  Copyright 2026 The InfiniFlow Authors. All Rights Reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

package orchestrator

import (
	"context"
	"fmt"
	"strings"

	"ragflow/internal/rag/advanced_rag/harness"
	"ragflow/internal/rag/prompts"
)

// Query rewriter (Phase 4 of the pipeline).
//
// Mirrors Python orchestrator/query_rewriter.py::rewrite_gap_to_query.
//

// RewriteDeps are the dependencies of RewriteGapToQuery.
type RewriteDeps struct {
	// Model generates the JSON object.
	Model JSONModel
	// Prompts supplies the "sca_query_rewrite" template.
	Prompts harness.PromptLoader
	// BridgeValues are already-resolved values across the claims (e.g. the
	// confirmed shows in a "which finale ran longest" question). The rewriter
	// anchors the new query to these, so a missing hop is re-searched with the
	// resolved upstream value instead of re-deriving it.
	BridgeValues []string
	// ResearchContext is the rendered block describing what previous rounds
	// already tried (queries + outcomes) and what the evidence pool contains.
	// It lets the rewriter aim at UNCOVERED angles instead of paraphrasing dead
	// ones.
	ResearchContext string
}

// RewriteGapToQuery mirrors Python rewrite_gap_to_query.
//
// gaps are (what, search_hint) tuples — the forward gaps the SCA identified.
//
// Returns [{"query": "..."}] — targeted, retrievable queries. Empty when the
// rewrite is unavailable or fails, in which case the caller falls back to the
// original gap text.
func RewriteGapToQuery(ctx context.Context, deps RewriteDeps, question string, gaps []MissingPiece) []map[string]string {
	if deps.Model == nil || len(gaps) == 0 {
		return nil
	}
	// Python wraps the call in @in_phase("rewrite").
	ctx, done := harness.Phase(ctx, harness.PhaseRewrite)
	defer done()

	var gapsLines []string
	for _, g := range gaps {
		gapsLines = append(gapsLines, fmt.Sprintf("- what: %s; hint: %s", g.What, g.SearchHint))
	}
	gapsText := strings.Join(gapsLines, "\n")

	var bridgeLines []string
	for _, b := range deps.BridgeValues {
		if strings.TrimSpace(b) != "" {
			bridgeLines = append(bridgeLines, "- "+b)
		}
	}
	bridgeText := strings.Join(bridgeLines, "\n")

	prompt := prompts.Render(deps.Prompts, "sca_query_rewrite", "", map[string]string{
		"question":         question,
		"gaps":             gapsText,
		"bridge_values":    bridgeText,
		"research_context": deps.ResearchContext,
	})

	result, err := deps.Model.GenJSON(ctx, prompt)
	if err != nil {
		_LOG.Printf("[QueryRewrite] failed: %v", err)
		return nil
	}
	data, ok := result.(map[string]any)
	if !ok {
		return nil
	}
	rawQueries, _ := data["queries"].([]any)

	var out []map[string]string
	for _, q := range rawQueries {
		var query string
		if m, ok := q.(map[string]any); ok {
			// Accept `query` (canonical) or `question` (observed drift).
			if v := m["query"]; v != nil {
				query = strings.TrimSpace(fmt.Sprint(v))
			} else if v := m["question"]; v != nil {
				query = strings.TrimSpace(fmt.Sprint(v))
			}
		} else if q != nil {
			query = strings.TrimSpace(fmt.Sprint(q))
		}
		if query == "" || containsQuery(out, query) {
			continue
		}
		out = append(out, map[string]string{"query": query})
	}
	return out
}

func containsQuery(queries []map[string]string, q string) bool {
	for _, item := range queries {
		if item["query"] == q {
			return true
		}
	}
	return false
}
